### PR TITLE
add-curl-advisories-importer

### LINKED ISSUE #1166: Add CURL advisories data source
See https://curl.haxx.se/docs/vulnerabilities.html and https://curl.se/docs/security.html
The actual sources are in https://github.com/curl/curl-www/tree/master/docs
And there are perl scripts to generate the tables.
DO NOT scrape the web pages. Use the repo and scripts instead.

### PR DESCRIPTION
**Pull request by ambuj-1211**
This PR fixes #1166. 
The test file is a dummy and will be completed once the curl.py importer is completed.

### CODE REVIEW

File: vulntotal/tests/test_osv.py [hunk 1]
```diff
@@ -73,4 +73,4 @@ def test_generate_payload_nuget_with_api_call(self):
         purl = PackageURL.from_string("pkg:nuget/moment.js@2.18.0")
         results = osv.generate_payload(purl)
         expected = {"package": {"ecosystem": "NuGet", "name": "Moment.js"}, "version": "2.18.0"}
-        assert results == expected
+        assert results == expected
```
Comment (TG1999) [inline vulntotal/tests/test_osv.py:76]: This change is not required

File: vulnerabilities/importers/curl.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import logging
```
Comment (TG1999) [inline vulnerabilities/importers/curl.py:1]: Add header

File: vulnerabilities/importers/curl.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+class CurlImporter(Importer):
+
+    spdx_license_expression = "MIT"
+    license_url = "https://github.com/curl/curl-www/blob/master/LICENSE"
+    repo_url = "https://github.com/curl/curl-www/"
+    importer_name = "Curl Importer"
+    api_url = "https://curl.se/docs/vuln.json"
+
+    def get_response(self):
```
Comment (TG1999) [inline vulnerabilities/importers/curl.py:28]: Please check utils.py and import `fetch_response` the function from there

File: vulnerabilities/importers/curl.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            return response.json()
+        raise Exception(
+            f"Failed to fetch data from {self.api_url} with status code: {response.status_code!r}"
+        )
+    
+    def advisory_data(self) -> Iterable[AdvisoryData]:
+
+        raw_data = self.get_response()
+        for data in raw_data:
+            cve_id = data["aliases"]
```
Comment (TG1999) [inline vulnerabilities/importers/curl.py:40]: What if "aliases" is not in data ? And will aliases be a string or a list ?
```suggestion
            cve_id = data.get("aliases") or ""
```

File: vulnerabilities/importers/curl.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            if not cve_id.startswith("CVE"):
+                logger.error(f"Invalid CVE ID: {cve_id} in package {data['database_specific']['package']}")
```
Comment (TG1999) [inline vulnerabilities/importers/curl.py:42]: Use get method to extract data from dictionary

File: vulnerabilities/importers/curl.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            yield parse_advisory_data(data)
+
+# Single dictionary is coming from the list of dictionaries in the input of below func.
+def parse_advisory_data(raw_data) -> AdvisoryData:
+
+    purl = PackageURL(type="curl", name="curl")
+    # add range of raw data accordingly as f string using first and last value of the list.
+    affected_version_range = NginxVersionRange.from_native(raw_data["vulnerable"])
+
+    fixed_version = SemverVersion(raw_data["affected"][0]["ranges"][0]["events"][1]["fixed"])
```
Comment (TG1999) [inline vulnerabilities/importers/curl.py:53]:
```suggestion
    fixed_version = SemverVersion(raw_data["affected"][0]["ranges"][0]["events"][1]["fixed"])
```
Check `get_item` method in utils.py and before accessing any index check the length of list

**Review by TG1999 — CHANGES_REQUESTED** [review]
@ambuj-1211 thanks ++ for your contribution, I have added some comments for your consideration. Also run `make valid` command for fixing formatting errors.

**Comment by TG1999** [comment]
@ambuj-1211 additionally please sign-off your commits

**Comment by ambuj-1211** [comment]
@TG1999  Apologies for the late response, I was stuck in my end sem exams. Done some changes as mentioned. I also want to know what to do for pURL, as in what should be the type, package name, and other components of the purl and also the version range for curl is not defined in univers what to do for that?

**Comment by TG1999** [comment]
type-generic namespace-domain name i.e curl.se name - name of the tool (curl, tiny-curl), version - as specified, download_url as qualifiers, so https://curl.se/download/curl-8.6.0.tar.gz this will be `pkg:generic/curl.se/curl@8.6.0?download_url=https://curl.se/download/curl-8.6.0.tar.gz`

**Comment by TG1999** [comment]
@ambuj-1211 try to collect all the versions of curl and commit a test to check with univers if all versions are effectively Semver Version

**Comment by ambuj-1211** [comment]
@TG1999 done the changes please review the commits

**Comment by pombredanne** [comment]
@ambuj-1211 hey, you are not planning to complete this after all?

**Comment by ambuj-1211** [comment]
I made a new pr regarding this issue therefore I closed this one, because there were some problems with that branch which I was not able to amend.

**Pull request closed without merging**